### ISSUE TITLE
Refactor TTS APIs into a general audio dispatching allowing different sources and sinks

### ISSUE THREAD

**Issue by kaikreuzer** [eclipse/smarthome#584]
migrated from Bugzilla [#463802](https://bugs.eclipse.org/bugs/show_bug.cgi?id=463802)
status NEW severity *enhancement* in component *Core* for *---*
Reported in version *unspecified* on platform *All*
Assigned to: Project Inbox

On 2015-04-02 09:11:10 -0400, Kai Kreuzer wrote:

> The current TTS implementations combine the creation of the audio stream with its output. E.g. the Google translate TTS service generates an mp3 and outputs it on the local soundcard.
> It would be much nicer, if the mp3 would be created by one service (the "source", here: Google Translate) and the output would be done by another one - this way, it could be played back as well on Sonos, Squeezebox etc. (the "sink").
> 
> When splitting it this way, it automatically means that it is not limited to TTS - but it could indeed be any kind of soundfile. So this refactoring should also address the "playSound" action at the same time.

**Comment by kdavis-mozilla** [comment]
I wanted to capture here several details related to this issue that Kai and our group discussed. I'll try to be pedantic to document the details of the discussion to make sure that all attendees agree on the discussion's results.

1. The current interface [`TTSService`](https://github.com/eclipse/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.multimedia/src/main/java/org/eclipse/smarthome/io/multimedia/tts/TTSService.java) should be refactored
2. The refactoring should introduce two new interfaces `AudioSource` and `AudioSink`, the names may change and are simply placeholders for now.
3. `AudioSource` is an interface any binding that is a sound source should implement. For example a binding for a microphone would implement this interface.
4. An `AudioSource` can be in any audio format. For example mp3, wav...
5. An `AudioSource` must allow a client of the interface to obtain its format. In other words it should expose sufficient get methods for a client to determine if it can play such an `AudioSource`.
6. An `AudioSource` can either be of a finite length, for example the mp3 of a song, or of an indefinite length, for example the live streaming of a speech.
7. `AudioSink` is an interface any binding that is a sound player should implement. For example a binding for a Sonos player should implement this interface.
8. The refactored [`TTSService`](https://github.com/eclipse/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.multimedia/src/main/java/org/eclipse/smarthome/io/multimedia/tts/TTSService.java) interface would redefine the `say` method to produce a `AudioSource`.
9. The `voice` parameter of the refactored [`TTSService`](https://github.com/eclipse/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.multimedia/src/main/java/org/eclipse/smarthome/io/multimedia/tts/TTSService.java) should uniquely identify the voice used in the say command. (If several TTS engines have voices of the same name name clashes must be dealt with somehow.)
10. A mechanism must be put in place to allow at runtime a  `AudioSource` to be dynamically paired with a compatible `AudioSink`. (This may be as simple as having an `AudioSink` expose a `bool isPlayable(AudioSource audioSource)` method then at runtime looping over all `AudioSink`'s looking for the first that returns `true` from its `isPlayable` method.)

If any details from the discussion were not captured here, please add them!

**Comment by kaikreuzer** [comment]
A few comments from my side:

- on 3 and 7: "binding... should implement": So far, the TTS implementations are not bindings, but an add-on type of its own. We will have `AudioSource`s and `AudioSink`s, which won't be related to bindings, but are simply local services (e.g. the system audio out). But it is right that microphones & speakers on "Things" will also serve as sources and sinks, so we will have to see how we can declare such functionalities within a Thing and how and when these are registered as a service.
- on 4: It should not be "any" format, but a format from a list that we should define
- on 8: Additionally, we should re-implement the [Audio action](https://github.com/kaikreuzer/openhab-core/blob/master/bundles/org.openhab.core.compat1x/src/main/java/org/openhab/io/multimedia/actions/Audio.java#L51) and move it to ESH, which would then bring together the right source+sink.
- on 10: Besides "auto-matching", we should also allow manually selection of sources and sinks (e.g. for the rule action, where the user might want to have different notifications on different speakers). In order to let users do this through UIs as well, we will have to think about adding some meta-data (labels, descriptions, etc.) as well.

**Comment by hkuhn42** [comment]
Hello,

i am working on  something  similar for some Time now and would like to contribut to this. I am currently on vacation without a computer but will be back tomorrow. The (very early) State of my implementation can be found at https://github.com/hkuhn42/sylvani but ist still very Rough...

**Comment by kaikreuzer** [comment]
Thanks @hkuhn42, I briefly browsed your repo and that looks indeed very similar to what our plans - so it would be great to have you on board here!

**Comment by kdavis-mozilla** [comment]
@hkuhn42 I've also taken a look. It does look very similar to what we had in mind. We should try to join forces!

**Comment by hkuhn42** [comment]
Ok, how do we go about this (sorry but as kai already learned i am new to this and also to github and may need some support)?
I could setup a repo with a fork and move the code there. Would that be a good way to get things started? 
If so, which package should i put it and which code should i add? Just the api? I also started to write basic implementations for the java sound api to support local audio devices but that api is not part of compact profile 2 so it violates the guidelines (as stated by the [guidelines](http://www.eclipse.org/smarthome/documentation/development/guidelines.html)).
 
In general the repo contains some experiments regarding voice control (text-to-speech and speech-to-text) as i did find the current support somewhat lacking. If anything of that code is interesting i can also move it over (i apologize for the missing comments and stuff ;) .

**Comment by kaikreuzer** [comment]
I think we should first start to roughly define the API (incl. package name etc.) according to the above requirements and agree on them.
@hkuhn42 Do you feel that your current code is already fulfilling all ideas above or does it actually address things that we might have forgotten and that should be considered?
Once we have agreement on this, you could create a fork of ESH, refactor your code if necessary and create a pull request against ESH, on which we can then continue any further discussion.

**Comment by hkuhn42** [comment]
I think that it covers most of the requirements as far as i understood them. I currently do not have a mechanism that matches the auto pairing as described in 10. My idea for that was to use a registry to register sources and outputs and also offer methods for simple playback and recording.
Also i have to take a look at that audio action to understand what it currently does.
For me it comes down to what would be more easy for you. You could have a look at the current state of the api or i can extract it into a fork first. If it does not match your requirements, we can always start from scratch. I added some more doc to hopefully make it easier for you to find the relevant parts in my repo. 
- The basic audio api is currently under https://github.com/hkuhn42/sylvani/tree/master/org.sylvani.io/src/main/java/org/sylvani/audio
- The TTS API is https://github.com/hkuhn42/sylvani/tree/master/org.sylvani.io/src/main/java/org/sylvani/io/voice
- And the command Interface is under https://github.com/hkuhn42/sylvani/tree/master/org.sylvani.io/src/main/java/org/sylvani/io/interpreter

**Comment by kdavis-mozilla** [comment]
Thinking a bit about the `ISyntheziserService` interface

```Java
public interface ISyntheziserService {

    public AudioSource synthesize(String text, Locale locale, String voice, AudioFormat requestedFormat)
            throws Exception;

}
```
A few things popped to mind

- A voice is, in all cases I am aware of, bound to a locale. In other, words specifying a `voice` implies a `locale`. So, I don't think `locale` is required as a parameter.
- If a GUI or voice interface wishes to display or say the available voices, it has no means to obtain the list of possible voices.
- It's unclear to me if `voice` also specifies the add-on from which the targeted voice comes. Two TTS add-ons may have voices with the same name. Hence, we should provide a means to disambiguate.
- Is audio transcoding the responsibility of each `ISyntheziserService` implementation or do they simply output one of the `n` formats they know of and throw otherwise.
- If audio transcoding is the responsibility of each `ISyntheziserService` implementation, I think it shouldn't be and we should introduce a new interface/implementation to do transcoding.

I haven't had time to look at the audio portion in detail. I'll comment on `ICommandInterpreter`in issue #1028

**Comment by hkuhn42** [comment]
> A voice is, in all cases I am aware of, bound to a locale. In other, words specifying a voice implies a locale. So, I don't think locale is required as a parameter.

That is actually a point i spend a lot of time thinking about and did not come up with an ideal solution. Voices do imply locale but when using the different services i experimented with, i usually did not care for the actual voice used but rather just used the first to match the my locale (due to the fact that apart from english, locale support gets thin very fast..). Also for some, like mary tts, additional voices have to be installed first.  So the two parameters were meant to be one or the other (which is not ideal). A way to tell the service to use any voice for a given locale would be a must have for me.
Regarding the actual voice selection i was torn between using strings for simplicity and an interface  which every service would have to implement. 

> Is audio transcoding the responsibility of each ISyntheziserService implementation or do they simply output one of the n formats they know of and throw otherwise.

My aproach was that the implementation should the default supported format (or one of them). Converting from one of these to another should than be part of an audio api if needed. However, audio transcoding is heavy stuff so my idea was to start with not doing any trancoding but rely on available formats.

So taking into account your comments and some of the my thoughts a new version of the interface could look like: 
```

/**
	 * Synthezise voice output for the given text in the given locale with the specified voice and format
	 * 
	 * Voice and format info should be optional
	 * 
	 * @param text
	 * @param voice
	 * @param requestedFormat
	 * @return
	 * @throws Exception
	 */
    public AudioSource synthesize(String text, Voice voice, AudioFormat requestedFormat)
            throws Exception;
    
    /**
     * Array containing all supported audio formats this SyntheziserService can produce
     *
     * @return an array of AudioFormat
     */
    public AudioFormat[] getSupportedFormats();
    
    /**
     * Get the default voice for the given locale or null if it is not supported 
     * 
     * @param locale
     * @return
     */
    public Voice getDefaultVoice(Locale locale);

    /**
     * An array containing all supported voices of this SyntheziserService
     * 
     * @return an array of Voice implementations
     */
    public Voice[] getAvailableVoices();
...

**Comment by kdavis-mozilla** [comment]
@hkuhn42 I'm liking the direction this is going!

I've a suggestion, motivated by the [WebSpeech API](https://dvcs.w3.org/hg/speech-api/raw-file/tip/webspeechapi.html), for the `Voice` interface that would let us get rid of the `getDefaultVoice()` method. If the `Voice` interface takes a form something like
```Java
public interface Voice {
       /**
        * Globally unique identifier of the voice
        *
        * @return A URI uniquely identifying the voice globally
        */
        public java.net.URI getVoiceURI();

       /**
        * Name of the voice, usually used for GUI's or VUI's
        *
        * @return Name of the voice, may not be globally unique
        */
        public String getName();

       /**
        * Locale of the voice
        *
        * @return Locale of the voice
        */
        public Locale getLocale();

       /**
        * Indicates if the voice requires internet access
        *
        * @return A boolean indicating if the voice in local to the hardware
        */
        public boolean isLocalVoice();
}
```
then sufficient information would be provided by the `getAvailableVoices()` method to obtain a `Voice` for a given `Locale`, which, as far as I understand, is your use case for `getDefaultVoice()`.

**Comment by hkuhn42** [comment]
@kdavis-mozilla the getDefaultVoice was meant to help in finding the most suitable voice for a given locale. However i dropped it as finding a voice for a locale can alwys be implemented outside of such a service.

Regarding the Voice interface, i see you point but i would rather not mix the "usage" interface with any kind of voice of configuration management as this seems to be highly implementation dependant. E.g. for online services, isLocale would have to always return false (so in my understanding the whole services allways depends on internet beeing available). In this case, i would rather raise an exception if no internet is available. 
For using an uri to identify a voice, i am not sure if this is the way its done in ESH. Beeing able to identify a voice is definitly something we should make available.
If nobody objects, I will modify my implementation accordingly later today and, if no more comments come in, move it to an ESH fork for further discussion.

@kaikreuzer should we make a new voice bundle or shoud this be a part of the existing multimedia bundle?

**Comment by kdavis-mozilla** [comment]
@kdavis-mozilla A few questions/comments...

- "...as finding a voice for a locale can alwys be implemented outside of such a service" How? If the service doesn't expose this information I don't see how this is done.
- A GUI/VUI can also use the TTS service to allow selection of possible voices. I see this as a valid usage. Thus, I see the `Voice` interface as valid interface to have.
- Why raise an exception instead of having `isLocalVoice()` return false? A non-local service is not an "abnormal condition".
- Currently ESH uses "the name of the voice". Normally voices are named things like "Fred", "Alice"... As such names are so generic there may exist name clashes between various `ISyntheziserService` implementations. The `URI` disambiguates this.  Maybe we can use something else. But a `URI` is a natural choice.

Before we start on implementation, we need to consider the `AudioFormat` and `AudioSource` interfaces in more detail.

**Comment by kaikreuzer** [comment]
> Currently ESH uses "the name of the voice".

We should probably rather talk about an id and a label - the id being the unique technical identifier, the label is the text to show in UIs. Usually, we have UI labels also localizable, but for voices this is imho not needed.
Btw, regarding "meta-information for UIs", we will probably have to define structures as well, e.g. stuff similar as for [bindings](https://www.eclipse.org/smarthome/documentation/development/bindings/xml-reference.html#example-1). But I think we can keep that for a later phase and concentrate on the "internals" for now. What can be easily done for the start though is adding a config description, which will make the service pop-up in the Paper UI. See [here](https://community.openhab.org/t/jdbc-persistence-heave-from-oh1-to-oh2-esh-choosing-a-namespace/7487/5?u=kai) for some hints on how to do this.

> isLocalVoice()

I am not sure if we would need this information - so far, we have nowhere an indication whether something is local or remote. And anyhow, I would assume that it usually applies to the whole TTS service and that it is a rare exception that a single TTS service has local AND remote voices.

> @kaikreuzer should we make a new voice bundle or shoud this be a part of the existing multimedia bundle?

We need to discuss the namespace for all the new components. I could imagine that we move this out of the "multimedia" package into e.g.
`org.eclipse.smarthome.io.voice.tts`, 
`org.eclipse.smarthome.io.voice.sst` and 
`org.eclipse.smarthome.io.voice.interpreter`. 
Any better suggestions?

**Comment by kdavis-mozilla** [comment]
@kaikreuzer If we want to call the `URI` the `id` and the `name` the `label`. I'm fine with that.

However, I think keeping the `id`a `URI` is a good idea as it leads to natural namespaces. The MaryTTS group cold name their voices something like `http://mary.dfki.de/voices/en/unitselection/Fred`, while the FreeTTS could use something like `http://freetts.sourceforge.net/Fred`, and the two naming schemes would never conflict.

As far as "meta-information for UIs" I agree. But I guess we should get the internals out of the way first.

The reason I thought `isLocalVoice()` would be useful is for security conscious users. They'd likely not want anything going out over the network.

As for namespaces, @kaikreuzer whatever you decide upon in fine.

**Comment by kaikreuzer** [comment]
@kdavis-mozilla Note that for an URI, only scheme and path are mandatory. So `mary:Fred` and `freetts:Fred` are valid URIs as well - and these would be much nicer to be used e.g. within a textual rule, where you want to pass them as a parameter like `say("mary:Fred", "Hello world!")`.
And again for this use case, I think using a simply String instead of the java URI class is easier to use. For such unique ids, we usually call the field then [uid](https://www.eclipse.org/smarthome/documentation/javadoc/org/eclipse/smarthome/automation/Rule.html#getUID()) , so this might be a good choice here as well. The JavaDoc can then say that it should have a "prefix:voicename" format.

**Comment by kdavis-mozilla** [comment]
@kaikreuzer We can use a `String` and impose validity through documentation, works for me.

On `isLocalVoice()` I guess the missing part for me was the statement...

> ...And then it can be nicely documented for the service itself...

If that's always the case, we don't need the `isLocalVoice()` method.

**Comment by kdavis-mozilla** [comment]
Before we jump into implementation, I think we still need to agree on the contents of [org.sylvani.audio](http://bit.ly/1VvogVK) or more specifically what our version of this package will be.

A quick summary of the things we've explicitly used, `AudioSource` and `AudioFormat`
```Java
public interface AudioSource {

    /**
     * Returns the human readable name of the source
     *
     * @return the human readable name of the source
     */
    public String getName();

    /**
     * an array containing all supported audio formats
     *
     * @return
     */
    public AudioFormat[] getSupportedFormats();

    /**
     * Gives access to an InputStream for reading audio data, the format is the default format
     *
     * @return InputStream for reading audio data
     * @throws AudioException
     */
    public InputStream getInputStream() throws AudioException;

    /**
     * An inputstream for reading audio data, the format is set to the given format
     *
     * @param format the desired format (one of getSupportedFormats) or null for the default format
     * @return InputStream for reading audio data
     * @throws AudioException
     */
    public InputStream getInputStream(AudioFormat format) throws AudioException;

    /**
     * Load data from this AudioSource to the given {@link AudioOutput} 
     *
     * @param output
     * @throws AudioException
     */
    public void stream(AudioOutput output) throws AudioException;
    
    /**
     * Returns true if this AudioSource can stream to the given {@link AudioOutput}
     * false otherwise
     * 
     * @param source
     * @return true if the AudioSource can be processes
     */
    public boolean canStream(AudioOutput source);
}

public class AudioFormat {

    private AudioCodec codec;
    private AudioContainer container;
    /**
     * bit depth (https://en.wikipedia.org/wiki/Audio_bit_depth)
     * or
     * bit rate (https://en.wikipedia.org/wiki/Bit_rate)
     * depending on codec
     */
    private int bits;
    /**
     * sample frequence
     */
    private long frequency;

    public AudioCodec getCodec() {
        return codec;
    }

    public void setCodec(AudioCodec codec) {
        this.codec = codec;
    }

    public AudioContainer getContainer() {
        return container;
    }

    public void setContainer(AudioContainer container) {
        this.container = container;
    }

    public int getBits() {
        return bits;
    }

    public void setBits(int bits) {
        this.bits = bits;
    }

    public long getFrequency() {
        return frequency;
    }

    public void setFrequency(long rate) {
        this.frequency = rate;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof AudioFormat) {
            AudioFormat format = (AudioFormat) obj;
            if (format.getCodec() != getCodec()) {
                return false;
            }
            if (format.getContainer() != getContainer()) {
                return false;
            }
            if (format.getBits() != getBits()) {
                return false;
            }
            if (format.getFrequency() != getFrequency()) {
                return false;
            }
            return true;
        }
        return super.equals(obj);
    }
}
```
Along with the things we've implicitly used, ignoring exceptions,

```Java
public interface AudioOutput {

    /**
     * Returns the human readable name of the source
     *
     * @return the human readable name of the source
     */
    public String getName();

    /**
     * Array containing all supported audio formats this output can process
     *
     * @return
     */
    public AudioFormat[] getSupportedFormats();

    /**
     * An output stream for writing audio data in the default {@link AudioFormat} of this output
     * 
     * @return an {@link OutputStream} 
     * @throws AudioException 
     */
    public OutputStream getOutputStream() throws AudioException;

    /**
     * An output stream for output audio, the format is set to the given format, throws and {@link UnsupportedAudioFormatException} if the given
     * format is not supported
     *
     * @param format the desired format (one of getSupportedFormats) or null for the default format
     * @return an OutputStream to read data from this output
     * @throws AudioException thrown among other reasons if the given format is not supported
     */
    public OutputStream getOutputStream(AudioFormat format) throws AudioException;

    /**
     * Process audio data from the provided {@link AudioSource} throws and {@link AudioException} if matching formats are found
     * 
     * @param source
     * @throws AudioException
     */
    public void stream(AudioSource source) throws AudioException;
    
    /**
     * Returns true if the given {@link AudioSource} can be processes by this output
     * 
     * @param source
     * @return true if the AudioSource can be processes
     */
    public boolean canStream(AudioSource source);
}

public enum AudioCodec {

    /**
     * PCM Signed
     * 
     * http://wiki.multimedia.cx/?title=PCM#PCM_Types
     */
    PCM_SIGNED

    ,
    /**
     * PCM Unsigned
     * 
     * http://wiki.multimedia.cx/?title=PCM#PCM_Types
     */
    PCM_UNSIGNED

    ,
    /**
     * MP3 Codec
     */
    MP3

    ,
    /**
     * Vorbis Codec
     */
    VORBIS
}

public enum AudioContainer {

    /**
     * NONE
     * AudioCodec encoded data without any container header or footer
     *
     * eg MP3 is a non container format
     */
    NONE

    /**
     * Microsofts wave container format
     * http://www.zytrax.com/tech/audio/formats.html#wav-format
     *
     * for a list of codesc supported by WAV see
     * http://www.opennetcf.com/library/sdf/html/60ca47dc-0b9d-2be4-a738-d0080c6fe10c.htm
     * 
     * the riff audio format
     */
    ,WAVE
    

    /**
     * http://www.xiph.org/ogg/
    */ 
    ,OGG   
}
```

**Comment by hkuhn42** [comment]
@kaikreuzer @kdavis-mozilla 
> If that's always the case, we don't need the isLocalVoice() method.

This would be a use case for an annotation: We should create two annotations that would mark the services as being @offline and @online. This way, this information could be used by humans and software alike). However this could possibly be a general functionality that imho would belong in some base package of ESH. Lucky enough we have exactly the right developer to ask at hand: @kaikreuzer what do you think?

About the audio package:

We obviously should change both

`public AudioFormat[] getSupportedFormats();`
to
`public Set<AudioFormat> getSupportedFormats();`

As stated somewhere in the javadoc, i ignored the Endian problem in AudioFormat. We probably should add something for that. A boolean would probably be just enough.

 and fix the typos in javadoc :smile:

**Comment by hkuhn42** [comment]
As the audio api can be used without TTS or STT (for example to support notification sounds) i suggest we move it to a seperate bundle. I created [org.eclipse.smarthome.io.audio](https://github.com/hkuhn42/smarthome/tree/audio/bundles/io/org.eclipse.smarthome.io.audio) for that purpose.

**Comment by kdavis-mozilla** [comment]
@kaikreuzer @hkuhn42 A naïve question, but why don't we simply use the standard java package [javax.sound.sampled](https://docs.oracle.com/javase/8/docs/api/javax/sound/sampled/package-summary.html)?

As far as I can see, `AudioSource` would correspond to [TargetDataLine](https://docs.oracle.com/javase/8/docs/api/javax/sound/sampled/TargetDataLine.html) and `AudioFormat` to [AudioFormat](https://docs.oracle.com/javase/8/docs/api/javax/sound/sampled/AudioFormat.html).

**Comment by hkuhn42** [comment]
@kdavis-mozilla I actually thought a lot about this before i wrote my own api definition.
I came up with a number of points. From my [audio readme](https://github.com/hkuhn42/sylvani/blob/master/org.sylvani.io/src/main/java/org/sylvani/audio/readme.md)
>There are multiple reasons for not using the Java Sound API:
>a) it is horrible for so many reasons (among them: nobody seems to know why target is for input and  source is for output and there is an audioInputStream but no AudioOutputStream) 
>b) it only defines constants and values for formats and codes it supports (and it only supports very basic ones)
>c) it is not part of compact profile 2 which is the target for eclipse smart home so any api using java sound can only be an extension 
>d) it does not offer a easy to use way to choose the output or input source / targets 
>e) java itself offers a second sound related api with the java media framework and both are not widely supported

**Comment by kdavis-mozilla** [comment]
@hkuhn42 Great! I just wanted to make sure we weren't overlooking the obvious.

**Comment by kaikreuzer** [comment]
> We should create two annotations that would mark the services as being @offline and @online

I still doubt that this is of much value - nobody yet ever asked for such a flag for any of the 120+ bindings; it is usually clear to them already what the binding is about and what the technology behind it does. If we really see the need for such formal annotations, we can still add it any time later.

> As the audio api can be used without TTS or STT (for example to support notification sounds) i suggest we move it to a seperate bundle.

Makes sense!

Regarding the interfaces: 

- I think we do not need both `AudioSource.stream` and `AudioOutput.stream`. Having the output stream a source should be enough, or am I missing something?
Likewise I would remove `AudioSource.canStream`.
- We should replace the `name` by a technical `id` and a human-readable `label`.
- What about allowing the source to offer different streams/content like with `public InputStream getInputStream(String contentId)`? We would also need additional methods through which an `AudioSource` can tell, which `contentId`s exist. But I am thinking e.g. of different radio stations, an mp3 library with 3 different door bells sounds etc. Or would you want to keep the content selection completely out of the technical interface? But if, what would be the other option? Having an `AudioSource` service registered for every single content option would definitely be too heavy.

**Comment by hkuhn42** [comment]
> I think we do not need both AudioSource.stream and AudioOutput.stream. Having the output stream a source should be enough, or am I missing something? Likewise I would remove AudioSource.canStream.
Having stream on both interfaces is just nice and symmetric but apart from that it could easily be removed and replaced with a comment to point this out.   

> We should replace the name by a technical id and a human-readable label.
Fine with me, if we do not have both (like with an actual file) they can easily be equal 

> Or would you want to keep the content selection completely out of the technical interface
Actually yes, i was always thinking about a kind of ContentProvider Interface that defines content selection. This could also offer named groups / folders and additional metadata about the provider.
That kind of functionality would probably not be needed for Services like STT or TTS that do not offer a selection. 

I wonder wether we should not separate the TTS and AudioFeature into two issues at this point?

@kdavis-mozilla I think we are now approaching the point where we should move the discussed TTS code into a bundle to make it easier to refine it.

**Comment by kaikreuzer** [comment]
> I wonder wether we should not separate the TTS and AudioFeature into two issues at this point?

Yes, this issue is primarily about "general audio dispatching allowing different sources and sinks" - so we can separate the discussion around TTS into a new issue.

**Comment by kdavis-mozilla** [comment]
@hkuhn42  "I think we are now approaching the point where we should move the discussed TTS code into a bundle to make it easier to refine it. What do you think?" Yes!

@kaikreuzer  @hkuhn42  Has the new audio issue been set up and I just "missed the boat"?

**Comment by kaikreuzer** [comment]
> Has the new audio issue been set up

You mean the TTS issue? No, feel free to do so!

**Comment by kdavis-mozilla** [comment]
@kaikreuzer I think I mean the "AudioFeature" issue. Explicitly, the design of the interfaces/classes `AudioSource`, `AudioFormat`... that we are talking about here, but might be better addressed in a separate issue.

**Comment by kaikreuzer** [comment]
As per my https://github.com/eclipse/smarthome/issues/584#issuecomment-187367915, I think this issue here is the right place to discuss the general audio stuff.

**Comment by kdavis-mozilla** [comment]
@kaikreuzer  Ok, but it seems slightly odd as most of this issue's discussion has been dealing with the TTS interfaces.

**Comment by kaikreuzer** [comment]
I wouldn't say so. 7 out of 10 points from https://github.com/eclipse/smarthome/issues/584#issuecomment-182771085 are about audio, not TTS.

**Comment by kdavis-mozilla** [comment]
Just to kick the discussion off again. Here are a few thoughts on the current audio API

- AudioSource.getInputStream() "Gives access to an InputStream for reading audio data, the format is the default format” However, there is no getter to obtain what the default format is.
- Why the naming "AudioSource/AudioOutput" the usual pairing for English is Source/Sink and Input/Output. For example consider the Java classes Pipe.SourceChannel/Pipe.SinkChannel and InputStream/OutputStream
- AudioOutput.getOutputStream() "Gives access to an OutputStream for writing audio data in the default AudioFormat of this output” However, there is no getter to obtain what the default format is.
- Should we limit the codecs with an enum? Adding new codecs then requires the enum changes and switch statements throughout our code and others code will have to change.
- Similar story for AudioContainer. Adding new containers with require changing the enum and switch statements throughout our code and other’s code.
- How do we want to solve the endian problem for AudioFormat. Simply a method AudioFormat.isBigEndian() that returns a boolean?
- I’d vote for either keeping both AudioSource.stream() and AudioOutput.stream() or getting rid of both. Keeping only one seems oddly asymmetric.
- I’d vote for keeping the content selection completely out of the technical interface. If we need it we can layer it on top of the technical interface later.

**Comment by kaikreuzer** [comment]
> Why the naming "AudioSource/AudioOutput"

as per the title of this issue, I vote for source/sink

> Should we limit the codecs with an enum?

Codecs/containers can imho be Strings - we should clearly list possible values in the documentation though.

> I’d vote for either keeping both AudioSource.stream() and AudioOutput.stream() or getting rid of both 

then I would remove both, as we have access to the input/output streams and the framework can do the hand-over.

> I’d vote for keeping the content selection completely out of the technical interface. 

Hm, ok, so far I have thought of AudioSources to be registered services in the system, that can be listed e.g. in a UI for selecting one - similar to the AudioSinks. Defining AudioSources now as use-once-and-throw away instances, makes it oddly asymmetric, doesn't it?

**Comment by hkuhn42** [comment]
First, i updated my [draft](https://github.com/hkuhn42/smarthome/tree/master/bundles/io/org.eclipse.smarthome.io.audio/src/main/java/org/eclipse/smarthome/io/audio) by renaming AudioOutput to AudioSink, removing all convenience methods and Changing the codecs/containers to strings according to the discussion. We can always add some of the again later on if the need arises. 

> > Should we limit the codecs with an enum?
> Codecs/containers can imho be Strings - we should clearly list possible values in the documentation though.

The reason for using enums was to prevent different implementations from using different names for the same formats codecs / containers. However i see your points. In my opinion, we should at least consider offering String constants for the most common formats. This way we can give some guidance without sacrificing flexibility.

> > I’d vote for keeping the content selection completely out of the technical interface.
> 
> Hm, ok, so far I have thought of AudioSources to be registered services in the system, that can be listed e.g. in a UI for selecting one - similar to the AudioSinks. Defining AudioSources now as use-once-and-throw away instances, makes it oddly asymmetric, doesn't it?

We definitely should thread both in the same way.
I thought a lot about this over the last couple of days and my understanding as of today is that AudioSink and AudioSource are a kind of handle to audio streams and blobs with additional format related information. 
I think that the way we plan to use the Source / Sinks in TTS/STT already shows that we will need some kind of providers with different apis. Streaming services or file stores will probably need support for some kind of groups and folders instead of locale or voice related config. Physical devices may need additional info about location or status. 
So i suggest we add AudioSink / AudioSource provider interfaces like STTService and TTSService and also move things like format negotiation into these (like already in place with those two). I think this will clear up the whole api and give us the opportunity to cater for special needs. 
So the io.audio bundle can define providers for a file system like structure to support playback of files or recording sounds and also for physical devices like speakers and mics as a base. The io.voice will define  providers for recognizing and synthesizing voice and so on. 
This way, if additional audio related add ons (like a phone or doorbell system) will be created, they can add additional methods  (like support for dialing or opening a door) for their special kind of content selection (with content selection meaning everything like choosing a radio station, a song, a voice, a format ...). 
So i also removed format options from the AudioSink and AudioSource interfaces from my draft.

I know that this would have a lot of impact on the whole issue but i also think it would be the better approach.

**Comment by kaikreuzer** [comment]
Treating source/sink instances only as a handle effectively means that we also do not need an id, a label and the audioformat (as it has already been negotiated upfront) on them. This leaves us with a simple `getInput/OutputStream`, doesn't it?

**Comment by kdavis-mozilla** [comment]
Just a few comments on your [draft](https://github.com/hkuhn42/smarthome/tree/master/bundles/io/org.eclipse.smarthome.io.audio/src/main/java/org/eclipse/smarthome/io/audio).

I like your take on [AudioCodec](https://github.com/hkuhn42/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.audio/src/main/java/org/eclipse/smarthome/io/audio/AudioCodec.java) and [AudioContainer](https://github.com/hkuhn42/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.audio/src/main/java/org/eclipse/smarthome/io/audio/AudioContainer.java) combining the advantages of an `enum` with that of a `String` by offering `String` constants for the most common formats.

Also, I'd guess there is a typo in the [AudioSink](https://github.com/hkuhn42/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.audio/src/main/java/org/eclipse/smarthome/io/audio/AudioSink.java) documentation
```Java
    /**
     * A set containing all supported audio formats this output can process
     *
     * @return
     */
     public AudioFormat getFormat();
```
I'd guess this is just a leftover and should be 

```Java
    /**
     * Returns the supported audio format of this AudioSink
     *
     * @return
     */
     public AudioFormat getFormat();
```
Similarly, I'd guess the documentation for [AudioSource](https://github.com/hkuhn42/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.audio/src/main/java/org/eclipse/smarthome/io/audio/AudioSource.java) should read
```Java
    /**
     * Returns the supported audio format of this AudioSource
     *
     * @return
     */
     public AudioFormat getFormat();
```

**Comment by kdavis-mozilla** [comment]
@hkuhn42 In regards to your [comments](https://github.com/eclipse/smarthome/issues/584#issuecomment-188970871).

To be sure that we are on the "same page", are you proposing introducing a [Service Provider Interface](https://en.wikipedia.org/wiki/Service_provider_interface) much like that provided by the [javax.sound.sampled.spi](http://docs.oracle.com/javase/8/docs/api/javax/sound/sampled/spi/package-summary.html) package?

Do you have a set of proposed interfaces? They don't have to be final. But, for me, I think they will give me a better picture of what your proposal encompasses.

**Comment by hkuhn42** [comment]
> Treating source/sink instances only as a handle effectively means that we also do not need an id, a label and the audioformat (as it has already been negotiated upfront) on them. This leaves us with a simple getInput/OutputStream, doesn't it?

@kaikreuzer Having access to a name / label / id (one would be suffice) as well as the format would enable decoupling the "provider" from a consumer so i think we should definitely keep these. I try to give an example to clarify what I mean: 
A TTSService Implementation would deliver a audioSource with a requested format. In order to play it to the user, we would pass it to a DeviceProvider that manages speaker / audioPlayers. The DeviceProvider would than check for a speaker / audioPlayer that can handle the format at hand. In this scenario we could pass the  source to the DeviceProvider without having to either pass the format along or have a central format management in place.

**Comment by hkuhn42** [comment]
@kdavis-mozilla 
- regarding the comments: yes this is definitely a leftover and you are right. 
- regarding the Providers i actually did not mean like SPIs. For me, the drafts of STTService and TTSService we discussed in the other issues already are such providers as they create  AudioSources from text / consume AudioSinks to produce text and have dedicated parameter for that functionality. An interface that deals with audio files or streams would imho need to offer a somewhat different api and my suggestion was merely to not try and go for one size fits all even if this is tempting but rather create a number of more specialized interfaces. A playback Service could e.g. have parameters for location and or name of the device or speaker but would not need voice info. 
So to give examples of the ideas of two other Providers i had in mind (just rough drafts of ideas really):
```
interface PlaybackService {
public void playback(AudioSource source, String location);
public Set<String> getLocations();
}

interface AudioRepository {
  public List<String> getCollections();
  public List<AudioSource> getAudioSource(String collection);
}

```

**Comment by kaikreuzer** [comment]
> @kaikreuzer Having access to a name / label / id (one would be suffice) as well as the format would enable decoupling the "provider" from a consumer

I think the decoupling is already done through having a class instance. Doesn't this already uniquely identify the content?

> A TTSService Implementation would deliver a audioSource with a requested format

Which means that the TTSService is the service displayed to the user and thus requires label, etc. and the audioSource is a concrete instance of a given spoken text.
But you are right, the AudioSource should still have the AudioFormat inside, so it is more than just an InputStream.

> we would pass it to a DeviceProvider that manages speaker / audioPlayers

With your examples, we are completely getting rid off the `AudioSink` interface... I wonder if we should not add it back by having the "hand-over" part clearly isolated in our source/sink concept and not requiring all kinds of different service interfaces for special use cases. How about:

```java
public interface PlaybackService {
  AudioSink getAudioSink(String location);
  Set<String> getLocations();
}

public interface AudioSink {
   Set<AudioFormat> getSupportedFormats();
   boolean process(AudioSource audioSource);
}
```
Or am I missing your recent idea of how the `AudioSink` interface is meant to be used?

**Comment by kdavis-mozilla** [comment]
@kaikreuzer In regards to

> I think the decoupling is already done through having a class instance. Doesn't this already uniquely identify the content?

I would also think that the decoupling is already done through having a class instance. @hkuhn42 are we missing something in this regard?

On the two points

- The `TTSService` is the service displayed to the user and thus requires label, etc. 
- The `AudioSource` should still have the `AudioFormat` inside

I agree with both.

As to the "hand-over" I see no problem with

```Java
public interface PlaybackService {
  AudioSink getAudioSink(String location);
  Set<String> getLocations();
}

public interface AudioSink {
   Set<AudioFormat> getSupportedFormats();
   boolean process(AudioSource audioSource);
}
```

**Comment by hkuhn42** [comment]
@kaikreuzer 

> Which means that the TTSService is the service displayed to the user and thus requires label, etc. and the audioSource is a concrete instance of a given spoken text.

You are right. However for certain use cases (like a kind of ringtones) it might be helpful to have some way of identifying the concrete AudioSource / AudioSink.

>With your examples, we are completely getting rid off the AudioSink interface... I wonder if we should not add it back by having the "hand-over" part clearly isolated in our source/sink concept and not requiring all kinds of different service interfaces for special use cases. 

Sorry, you are right again, i was taken a bit off track there...

> Or am I missing your recent idea of how the AudioSink interface is meant to be used?

No not at all, imho  there is only one thing missing: i would add a way to access an outputStream as a convenience to enable users of PlaybackService to not have to create an AudioSource first if they choose not to.

So in slight variation:

```
public interface PlaybackService {
  AudioSink getAudioSink(String location);
  Set<String> getLocations();
}

public interface AudioSink {
   Set<AudioFormat> getSupportedFormats();
   boolean process(AudioSource audioSource);
   OutputStream getOutputStream();
}
```
If you both think this unnecessary we can leave the method out.

If you both are happy with this interface and you do not object, i will try to incorporate these changes in my fork and prepare a pull request in the evening ( @kaikreuzer:  i fear will need your help with the details)

**Comment by kdavis-mozilla** [comment]
@hkuhn42 I think I'm fine with the current interface

**Comment by kaikreuzer** [comment]
> However for certain use cases (like a kind of ringtones) it might be helpful to have some way of identifying the concrete AudioSource / AudioSink.

Let's only add it when we really need it. Until then, you can use toString() to have a string representing the instance :-)

> If you both think this unnecessary we can leave the method out.
> If you both are happy with this interface and you do not object

You missed the situation that one of us agrees and the other objects :sunglasses: 
The problem that I am seeing with `getOutputStream()`is that a sink might need to provide a specific output stream depending on the audio format. This method now assumes that the stream can directly deal with any content. So to be clean, we would probably rather have to have a `getOutputStream(AudioFormat format)`. Another problem is multi-threading. Assuming the sink is something like a speaker, there can only be a single source stream being processed at a time. The `process` method supports this nicely as it can either block until the previous stream is done or it can internally queue the request. Reaching out the output stream to an unknown number of callers can result in pretty bad concurrency problems and unexpected behavior.

**Comment by kdavis-mozilla** [comment]
@kaikreuzer  Good catch. I think you're right `getOutputStream()` would have to take an `AudioFormat` 
```Java
public interface AudioSink {
   Set<AudioFormat> getSupportedFormats();
   boolean process(AudioSource audioSource);
   OutputStream getOutputStream(AudioFormat audioFormat);
}
```
The `AudioFormat` was there in a previous revision but then somehow got lost in the copy and paste.

As to threading and the use of an `OutputStream`, I guess there are two cases we have to deal with:

1. An `OutputStream` instance from a given `AudioSink` instance is passed from one thread to another and each thread writes to it.
2. Several threads have `OutputStream` instances all from a single `AudioSink` instance and each thread writes to their own `OutputStream` instance.

For the first case I don't think there is anything special to do. The threads collectively "know" what they are doing as they pass around the single `OutputStream` instance.

For the second case I also don't see a "problem". I would just expect the  `AudioSink`, if it's a Sonos player say, to play all the audio from the different threads at the same time.

The question is: What if a given `AudioSink` can only accept input from a single `OutputStream` at a time? If this were the case, then it should be up to the `AudioSink` implementation to enforce this.

For example an `AudioSink` could have a `WeakReference<OutputStream>` member to *the*  `OutputStream`, called  `outputStreamWeakReference` say, and throw an exception when  `getOutputStream()` is called and `outputStreamWeakReference.get()` is not `null`.

**Comment by kaikreuzer** [comment]
>  I would just expect the AudioSink, if it's a Sonos player say, to play all the audio from the different threads at the same time.

My assumption is that this is not possible in 95% of all cases. So the sink can only provide one OutputStream instance at a time and has to make sure that it is only used by one thread.

> and throw an exception when getOutputStream() is called 

This is pretty ugly for the caller and the caller would have to retry by himself to find out when the stream might be available again - this is horrifying API design. I would prefer to remove this method and rediscuss, if we really come across a situation where it is desirable to directly obtain the raw output stream.

**Comment by kdavis-mozilla** [comment]
@kaikreuzer 

> I would prefer to remove this method and rediscuss, if we really come across a situation where it is desirable to directly obtain the raw output stream.

Sounds fine to me.

**Comment by hkuhn42** [comment]
In general i think that the AudioSource and AudioSink are now quite different in the way they work. 

However I agree with @kaikreuzer that we should just go ahead and see how the api works  out when used. In my experience you learn most about an api when when implementing and using it.

**Comment by hkuhn42** [comment]
@kaikreuzer i tried to create a pull request. Please have a look. Thanks!

**Comment by kaikreuzer** [comment]
Implemented APIs with https://github.com/eclipse/smarthome/pull/1132.
Now I guess we need to adapt implementations/framework as well for it.

@kdavis-mozilla You mentioned that you were porting `TTSServiceMacOS` - will you create a PR for it? What else do we need?

**Comment by kdavis-mozilla** [comment]
On our fork we've implemented

- [TTSService interfaces](https://github.com/mozilla/smarthome/commit/26340cf54013728a6fed87ca2a0e4b2349b4e720)
- [TTSService implementation](https://github.com/mozilla/smarthome/commit/26340cf54013728a6fed87ca2a0e4b2349b4e720) for OS X's `say` command
- [TTSService implementation](https://github.com/mozilla/smarthome/commit/adb179b60a09e505969408c7a2d2ff7eb900d0bd) for MaryTTS
- [STTService interfaces](https://github.com/mozilla/smarthome/commit/5effc03cc284789786969f2c3b06c5a1b9b8f5b6)
- [STTService implementation](https://github.com/mozilla/smarthome/commit/de54da6326c1b3ede2cd76c60cce74d0de9f7d73) for Kaldi
- [STT and TTS tests](https://github.com/mozilla/smarthome/commit/6cd55fd321063f381f33b614e2614814a57ec4c8)
- [Keyword spotting interfaces](https://github.com/mozilla/smarthome/commit/34f18f9766862acd29b8e231313b82d47095891f)
- An [orchestrator](https://github.com/mozilla/smarthome/blob/f1992ef953afd90fcba42eb9157ec33da37ca0aa/bundles/io/org.eclipse.smarthome.io.voice/src/main/java/org/eclipse/smarthome/io/voice/internal/VaaniOrchestrator.java) that coordinates the actions of these implementations  

But, we based all of this off our implementation of the [Audio interfaces](https://github.com/mozilla/smarthome/commit/c1eaddc718a8962b93e5e3fa17d2eb209437879e) as we didn't have time to wait for this pull request.

However, our implementation of the [Audio interfaces](https://github.com/mozilla/smarthome/commit/c1eaddc718a8962b93e5e3fa17d2eb209437879e) should be the same as the pull request here, mod whitespace issues and the like.

**Comment by kaikreuzer** [comment]
> we didn't have time to wait for this pull request.

Then it is probably now a good time to reconcile things - you should avoid having your fork differing too much from the ESH master. Hope you'll find the time to do PRs for your work!

**Comment by kgoderis** [comment]
@kdavis-mozilla (@kaikreuzer) Cfr #1200, where are you standing?

**Comment by kaikreuzer** [comment]
@kgoderis Unfortunately @kdavis-mozilla and team stopped working on these features.
I am currently trying to reconcile their state of work, clean it up and create a PR from it - I hope to have that available by the end of this week and I hope @hkuhn42 will be able to help reviewing and commenting it then.

**Comment by kgoderis** [comment]
@kaikreuzer playSound is M.I.A. ? cfr the initial post in this thread

**Comment by kaikreuzer** [comment]
Well, at the moment it is still only available in the [compat layer](https://github.com/kaikreuzer/openhab-core/blob/master/bundles/org.openhab.core.compat1x/src/main/java/org/openhab/io/multimedia/actions/Audio.java#L59), but I plan to migrate this to the new framework (and have it moved to ESH).

My biggest issue there is that [the plain javasound sink](https://github.com/eclipse/smarthome/tree/master/extensions/io/org.eclipse.smarthome.io.javasound) does not support mp3s... Therefore I plan to implement an "enhanced" sink in OH2, which uses the jl library, which is used by the "old" Audio action.
And as long as there is no sink that can play mp3s, I didn't bother to migrate the playSound action.

**Comment by kgoderis** [comment]
I need it for Sonos in my setup so I will give it a go

**Comment by kaikreuzer** [comment]
This is now fully in place (through quite a few PRs in the last two weeks, which I don't want to list all here).
